### ISSUE TITLE
issue about quota 

### ISSUE THREAD

**Issue by fchioralia** [lizardfs/lizardfs#625]
When using the quota in lizardfs, the limit where the quota is enforced is represented by the size of  chunks used (the size parameter from dirinfo).  
Bellow I have an example, where quota is exceeded, because there are many small files, but the actual size is half from what is intended to be the limit. 

> **lizardfs dirinfo  /home/31615**
   /home/31615:
   inodes:                      39683
   directories:                 6588
   files:                      33095
   chunks:                      33012
   length:                  745155290
   size:                   2911244288
   realsize:               5822488576 
> 
> **lizardfs repquota -u 31615 /home** 
>User       31615 +-           2911244288           1677721600           1887436800                39688                    0                    0
>

The increase of the quota limit is not an option. If instead of small files, will be only big files, the value represented by the actual size(length), will be close to the value of size.
Below an example, where exists only big files and the length is close to size.
> **lizardfs dirinfo /home/31086**
/home/31086:
 inodes:                         16
  directories:                    1
  files:                         15
 chunks:                         15
 length:                  760566386
 size:                    761277440
 realsize:               1522554880
>
> **lizardfs repquota -u  31086 /home**
User       31086 +-            761277440            524288000            734003200                   21                    0                    0
>

I wonder, if exists a workaround for using, as limit, the length parameter from dirinfo,  if not, 
it is possible to implement a parameter, that will specify which size to be used for quota, like: size set as default, length or realsize?

**Comment by psarna** [comment]
No, there's no workaround for that. Changing it would require rewriting some of our current quota implementation. Length parameter is not a good determinant of space used, because sparse files have a huge length while having almost no size. With current implementation we sometimes overestimate used size, but it still works well enough with sparse files.

**Comment by fchioralia** [comment]
When having a file server, where users should have standard quotas, this overestimation limit the actual size used and quota is not relevant anymore.
Because this is not an option for me, the solution for my problem, could be increasing users quota to the maximum of smallest files that could fill the length. If in directory will be normal files, then the length used  will be  bigger than the intended quota, but when there are only small files the restriction will appear OK. 
So if I want that restriction to be at the end at the length of 1000Gb, what should be the quota if every file is the smallest?  Can be made an estimation let's say: **quota = 1000Gb X smallest_file_multiplying_factor** ?

**Comment by psarna** [comment]
I'm not sure if I follow what the question is, but maybe this info would help:

Each nonempty file takes up at least 1 block of space, and our block size is 64kB. So, I suppose smallest_file_multiplying_factor would be 64kB in your calculations.

**Comment by fchioralia** [comment]
I've created a file with 1 byte length and the size returned is close to 64kB, so this it is.
Let's suppose, that I set 1GB quota to an user, for emails and absurdly, all emails have 10kB. So instead of 1GB of storage, intended to use, at 160MB length (1GB/6.4), the quota will be reached. I can see a very unhappy user.  
I know that lizard is supposed to be used mainly for big data, but in this particular case, quota lose it's relevance and can't be used.  I hope that you will reconsider changing, a little bit, this quota implementation, to satisfy all the needs :)

**Comment by psarna** [comment]
The point is that actual file size depends on the file systems you use in your chunkservers. Some of them may use only one 512B block for 1 byte file, but others can use up to 64kB, especially if they do not support sparse files. So, 64kB is the upper limit of what a small file could potentially take from real available space in your installation. Hence, this upper value is taken into consideration when counting quotas. Otherwise users could potentially use more real space than allowed by their quota.

You need to take into account that even a 1 byte file is presented in LizardFS as a 1-block file, and this 1 block is stored on a chunkserver. Now, chunkservers may use totally different file systems underneath, so it's hard to pinpoint the actual space your file needs, especially from master's perspective (which doesn't store any information about which file systems chunkservers use and what's their block size - that would be way too much unnecessary fuss).

To sum up, I think current quota counting mechanism is the most safe one, so it should stay. If other users have different opinions, then please share.

**Comment by 4Dolio** [comment]
Our imap data still loves on a native zfs filesystem.  And our imap service has it's own quota mechanics.  You may want to let your mail services handle quotas rather than the lfs filesystem.

In addition, I used to backup our imap to lfs natively.  However the tiny file sized quickly bloat the lfs metadata size and consume far too much RAM as a result.  I have since resorted to doing these backups into a loopback filesystem which is then backed by a single massive lfs sparse file.  This worked much better than the original native lfs method.  I have since taken the loopback method a step further and setup an iSCSI service backed by lfs. The iSCSI is now attached to our Xen as an SR on which I create a Virtual Disk attached to a VM and am now backing up our imap into a ZFS+compression backed by the iSCSI SR backed by lfs.  This latest method has also proven to work very well for us.  I can execute lfs master transitions and the iSCSI simply follows the master floating IP without missing a beat.

Just some food for thought for you to consider.

**Comment by guestisp** [comment]
I agree with 4Dolio
Storing millions of chunks on LFS seems to expose more trouble than using big files, in example, you can't scrub files in less than many month and rebalance is very very slow. You have to tune some settings, that will blow down the over all performance (but speed up replication a lot).

**Comment by 4Dolio** [comment]
Yes.. Unfortunately, Lizard does not do well with bazillions of teeny tiny objects.  I have often wondered if perhaps we could somehow use a directory inode as a pointer to a large single chunk file and store tons of little objects within the "directory chunk".... Such a feature might allow for lfs to excel at this use case where it currently does not perform well.  It could be a new magic property given to directory objects.. Or perhaps an automatic storage method where by any object < 64k is stored as a subset of the parent folder rather than as a distinct object of it's own... Anyway... For now, I have resorted to large Loopbacks for many old static archive and massive quantities of tiny object storage... The problem with these solutions is that they are not longer accessible by multiple native lfs clients, which is sort of a bummer, but I am able to deal with that trade off.   In the end, it all comes down to this.  I have 256GB(*2) of Ram available for all master hosts, so nothing gets to 'waste' any RAM on metadata unnecessarily or without good justification.  Given that, I still feel LizardFS is the best of breed for open source tunable redundant horizontally scalable massive storage clusters.  I can't wait to be able to say we have X PetaBytes, I think we are only at about 0.25-ish PB atm..

**Comment by guestisp** [comment]
Another solution would be to run a VM with NFS server on top of Lizard, but you'll add another abstraction layer that will kill performance.

Or exporting a *huge* file via iSCSI, but loopbacks are slow....

**Comment by 4Dolio** [comment]
yup.. these are some of the trade offs..

**Comment by fchioralia** [comment]
The performance is the main reason, why I consider using Lizard FS natively for storing files. Maybe when lizard will have a NFS server integrated, that will not impact the performance too much, then I will consider using NFS. 
Right now for my needs(a file server cluster for a few thousand users, 2TB tops), this is the best option. I observed that the density of small files(<64kB) is less than 2%, so to cover the situations when quota is reached faster, because of this files, I've multiplied the intended quota with 10, to cover most of the unpleasant situations. The users now will have more storage, than expected, but I don't think this will be a problem. 
I hope that, when the NFS server will be integrated, using the real size of files(length) as quota, will be possible.

**Comment by pkali** [comment]
In my opinion it is better to have quotas based on the true disk usage, not file lengths, so I would not change the current quota system.
I do understand your point though and this is amongst the most valuable input from community. We literally crave for such comments.
We are now working hard on delivering the new release with, among others, aforementioned (p)NFS support, with important media performance enhancements development following. Therefore we can not focus on small files at the moment. Our team does have breakthrough ideas on improving small files performance, but this must be postponed.
